### PR TITLE
Vectorize

### PR DESCRIPTION
**Pull request by smao-astro**
vectorize the following methods:
- on_boundary
- boundary_normal
- periodic_point
- inside

### CODE REVIEW

File: deepxde/geometry/geometry_nd.py [hunk 1]
```diff
@@ -91,28 +95,31 @@ def __init__(self, center, radius):
...
         ad = np.dot(xc, dirn)
-        return -ad + (ad ** 2 - np.dot(xc, xc) + self._r2) ** 0.5
+        return -ad + (ad ** 2 - np.sum(xc * xc, axis=-1) + self._r2) ** 0.5
 
     def distance2boundary(self, x, dirn):
         return self.distance2boundary_unitdirn(x, dirn / np.linalg.norm(dirn))
 
     def mindist2boundary(self, x):
-        return np.amin(self.radius - np.linalg.norm(x - self.center, axis=1))
+        return np.amin(self.radius - np.linalg.norm(x - self.center, axis=-1))
```
Comment (lululxvi) [inline deepxde/geometry/geometry_nd.py:116]: This is the same.

**Comment by lululxvi** [comment]
Lots of changes in this pull request:
- bc
    - on_boundary
- geom
    - inside
    - on_boundary
    - on_initial
    - boundary_normal
    - periodic_point
    - distance2boundary
    - distance2boundary_unitdirn
    - is_left
- PointSet
    - inside
    - values_to_func
- Polygon
    - random_points

Most changes are API changes. I think it is better to have a few pull requests step by step to make the modification more clear.

0. If the array is 2d, then `axis=1` and `axis=-1` are the same.
1. Here are 2 changes that does not change the logic: `PointSet.values_to_func` and `Polygon.random_points`. The first pull request can only vectorize these two.
2. Because `bc.on_boundary` is implemented by users, it is better to have backward compatibility. Assume `bc.on_boundary` is not vectorized, then vectorize other codes as more as possible.
3. Still assume `bc.on_boundary` is not vectorized, and then we write a wrapper to make `bc.on_boundary` vectorized in `BC.__init__()`, similar to `np.vectorize`. Then we can vectorize all codes.
4. Support both vectorized or not-vectorized `bc.on_boundary`. If not, then apply the wrapper in Step 3.

**Comment by smao-astro** [comment]
> Because bc.on_boundary is implemented by users, it is better to have backward compatibility. Assume bc.on_boundary is not vectorized, then vectorize other codes as more as possible.
> Still assume bc.on_boundary is not vectorized, and then we write a wrapper to make bc.on_boundary vectorized in BC.__init__(), similar to np.vectorize. Then we can vectorize all codes.

Explored a bit and still not clear how to do this, if we assume the user-defined `bc.on_boundary` apply on single points, and try to do a wrapper to vectorize, my first impression is that the program won't benefit from the acceleration provided by `numpy`'s c API (not an expert on how numpy works, correct me if I am wrong), which is one of the key points of this pull request.

**Comment by lululxvi** [comment]
In BC, change the following
```python
self.on_boundary = on_boundary
```
to
```
self.on_boundary = lambda x, on: np.array([on_boundary(x[i], on[i]) for i in range(len(x))])
```
In this way, `self.on_boundary` is vectorized, and we can vectorize our other codes. Overall, there are benefits.

**Comment by lululxvi** [comment]
Why `on_boundary` and `inside` should be N*1 array? Should we `keepdims`?

**Comment by smao-astro** [comment]
> Why `on_boundary` and `inside` should be N*1 array? Should we just a list or a 1D array?

Hi Lu, that is my implicit convention: all the array should be in shape of (N_sample, N_dimensions), so that all the operation and arrays work together consistently, maybe I should state so explicitly.

**Comment by lululxvi** [comment]
Yes, for all the points, it should be (N_sample, N_dimensions). But `on_boundary` and `inside` are bool arrays, a flatten array seems more intuitive? Because it doesn't have any extra dimenstion. Also, consider the following code
```python
x = np.array([[1], [2], [3]])
x[np.array([True, False, Ture])]  # It is array([[1], [3]])
x[np.array([[True], [False], [Ture]])]  # It is array([1, 3])
```
So `np.array([True, False, Ture])` seems better. What do you think?

**Comment by lululxvi** [comment]
By the way, could you make a separate pull request first to vectorize `Polygon.random_points`?

**Comment by smao-astro** [comment]
Sounds good, implemented.

File: deepxde/geometry/csg.py [hunk 2]
```diff
@@ -29,19 +29,22 @@ def __init__(self, geom1, geom2):
...
-        if self.geom2.on_boundary(x) and not self.geom1.inside(x):
-            return self.geom2.boundary_normal(x)
-        return np.zeros(self.dim)
+        return np.logical_and(
+            self.geom1.on_boundary(x), ~self.geom2.inside(x)
+        ) * self.geom1.boundary_normal(x) + np.logical_and(
+            self.geom2.on_boundary(x), ~self.geom1.inside(x)
+        ) * self.geom2.boundary_normal(
+            x
+        )
```
Comment (lululxvi) [inline deepxde/geometry/csg.py:47]: Seems a bug. 
```python
        return np.logical_and(
            self.geom1.on_boundary(x), ~self.geom2.inside(x)
        )[:, None] * self.geom1.boundary_normal(x) + np.logical_and(
            self.geom2.on_boundary(x), ~self.geom1.inside(x)
        )[:, None] * self.geom2.boundary_normal(
            x
        )
```

File: deepxde/geometry/geometry_2d.py [hunk 3]
```diff
@@ -382,39 +385,53 @@ def wn_PnPoly(P, V):
...
+                    np.hstack(
+                        [
+                            V[i, 1] > P[:, 1:2],
+                            V[i + 1, 1] <= P[:, 1:2],
+                            is_left(V[i], V[i + 1], P) < 0,
+                        ]
+                    ),
+                    axis=-1,
+                )
+                wn[tmp] -= 1
```
Comment (lululxvi) [inline deepxde/geometry/geometry_2d.py:414]: It is better to add the comments back.

File: deepxde/geometry/timedomain.py [hunk 4]
```diff
@@ -17,7 +17,7 @@ def __init__(self, t0, t1):
         self.t1 = t1
 
     def on_initial(self, t):
-        return np.isclose(t[0], self.t0)
+        return np.isclose(t, self.t0)
```
Comment (lululxvi) [inline deepxde/geometry/timedomain.py:20]: Use 1d array.

File: deepxde/initial_condition.py [hunk 5]
```diff
@@ -16,8 +16,7 @@ def __init__(self, geom, func, on_initial, component=0):
         self.component = component
 
     def filter(self, X):
-        X = np.array([x for x in X if self.on_initial(x, self.geom.on_initial(x))])
-        return X if len(X) > 0 else np.empty((0, self.geom.dim))
+        return X[self.on_initial(X, self.geom.on_initial(X)).flatten()]
```
Comment (lululxvi) [inline deepxde/initial_condition.py:19]: - Similar to `on_boundary`, in line 15
```python
self.on_initial = lambda ...  # return a 1d array
```
- Then no `flatten` here

**Comment by lululxvi** [comment]
There are still a few bugs, see above. Also, `inside` is vectorized, and then the following should be modified:
- `CSGUnion.random_points()`
- `CSGUnion.random_boundary_points`
- `CSGDifference.random_points()`
- `CSGDifference.random_boundary_points()`
- `CSGIntersection.random_points()`
- `CSGIntersection.random_boundary_points()`

**Pull request merged**